src/rgw/rgw_data_sync.cc

```diff
@@ -1413,6 +1413,202 @@ class RGWDataSyncSingleEntryCR : public RGWCoroutine {
+      yield call(new RGWReadRemoteBucketIndexLogInfoCR(sc, source_bs.bucket, &remote_info));
+      if (retcode < 0) {
+        tn->log(10, SSTR("full sync: failed to read remote bucket info. Writing "
+                        << source_bs.shard_id << " to error repo for retry"));
+        yield call(rgw::error_repo::write_cr(sync_env->store->svc()->rados, error_repo,
+                                            rgw::error_repo::encode_key(source_bs, std::nullopt),
+                                            timestamp));
+        if (retcode < 0) {
+          tn->log(0, SSTR("ERROR: failed to log " << source_bs.shard_id << " in error repo: retcode=" << retcode));
+        }
+        return set_cr_error(retcode);
+      }
+
+      //wait to sync the first shard of the oldest generation and then sync all other shards.
+      //if any of the operations fail at any time, write them into error repo for later retry.
+
+      source_bs.shard_id = 0;
+      yield call(sync_single_entry(sc, source_bs, remote_info.oldest_gen, key, timestamp,
+                                  lease_cr, bucket_shard_cache, marker_tracker, error_repo, tn, false));
+      if (retcode < 0) {
+        tn->log(10, SSTR("full sync: failed to sync " << source_bs.shard_id << " of gen "
+                        << remote_info.oldest_gen << ". Writing to error repo for retry"));
+        yield call(rgw::error_repo::write_cr(sync_env->store->svc()->rados, error_repo,
+                                            rgw::error_repo::encode_key(source_bs, remote_info.oldest_gen),
+                                            timestamp));
+        if (retcode < 0) {
+          tn->log(0, SSTR("ERROR: failed to write " << remote_info.oldest_gen << ":" << source_bs.shard_id
+                         << " in error repo: retcode=" << retcode));
+          return set_cr_error(retcode);
+        }
+      }
+      each = remote_info.generations.begin();
+      for (; each != remote_info.generations.end(); each++) {
+        for (sid = 0; sid < each->num_shards; sid++) {
+          source_bs.shard_id = sid;
+          tn->log(10, SSTR("full sync: syncing shard_id " << sid << " of gen " << each->gen));
+          yield_spawn_window(sync_single_entry(sc, source_bs, each->gen, key, timestamp,
+                            lease_cr, bucket_shard_cache, marker_tracker, error_repo, tn, false),
+                            cct->_conf->rgw_data_sync_spawn_window,
+                            [&](uint64_t stack_id, int ret) {
+                            if (ret < 0) {
+                              sid = source_bs.shard_id;
+                              for (; sid < each->num_shards; sid++) {
+                                source_bs.shard_id = sid;
+                                spawn(rgw::error_repo::write_cr(sync_env->store->svc()->rados, error_repo,
+                                        rgw::error_repo::encode_key(source_bs, each->gen),
+                                        timestamp), false);
```

Comment:
ok, this all makes sense. but i think it'd be cleaner to move the error handling out of the lambda and into the nested-for-loop. each iteration in the loop would check `if (retcode < 0)` to decide whether to spawn `rgw::error_repo::write_cr()` or `sync_single_entry()` on the gen/shard. that way you can keep using `yield_spawn_window()` for all of the `rgw::error_repo::write_cr()`s too:
```c++
      each = remote_info.generations.begin();
      for (; each != remote_info.generations.end(); each++) {
        for (sid = 0; sid < each->num_shards; sid++) {
          source_bs.shard_id = sid;
          if (retcode < 0) {
            yield_spawn_window(rgw::error_repo::write_cr(sync_env->store->svc()->rados, error_repo,
                                                         rgw::error_repo::encode_key(source_bs, each->gen),
                                                         timestamp),
                               cct->_conf->rgw_data_sync_spawn_window, std::nullopt);
          } else {
            tn->log(10, SSTR("full sync: syncing shard_id " << sid << " of gen " << each->gen));
            yield_spawn_window(sync_single_entry(sc, source_bs, each->gen, key, timestamp, lease_cr,
                                                 bucket_shard_cache, marker_tracker, error_repo, tn, false),
                               cct->_conf->rgw_data_sync_spawn_window,
                               [&] (uint64_t, int ret) { if (ret < 0) { retcode = ret; } });
          }
```